File: 921-Minimum-Add-to-Make-Parentheses-Valid.js
var minAddToMakeValid = function(S) {
    let paren = [...S],
        needed = []
    for (let i = 0; i < paren.length; i++) {
        if (S[i] === "(") {
            needed.push(S[i])
        } else if (needed[needed.length - 1] === "(") {
            needed.pop()
        } else {
            needed.push(S[i])
        }
    }
    return needed.length
};

File: 921MinimumAddtoMakeParenthesesValid.py
class Solution:
    def minAddToMakeValid(self, s: str) -> int:
        stack = []  
        for p in s:
            if p == "(":
                stack.append(p)
            else:
                tail = stack[-1] if len(stack) > 0 else ""
                if tail == "(":
                    stack.pop(-1)
                else:
                    stack.append(p)
        return len(stack)
        
